File: packages/python/src/toolbox/schema.py
"""Schema formatting — convert JSON Schemas to compact LLM-friendly formats.

Opt-in: ``pip install toolbox-gateway[schema]``

This module provides two key utilities:

1. ``schema_to_csv`` — Convert a JSON Schema to a compact CSV header with type hints
2. ``schema_to_markdown`` — Convert a JSON Schema to markdown field documentation

Both are designed for the ``toolbox explain`` command to return compact,
LLM-friendly tool documentation instead of raw JSON Schema blobs.
"""

from __future__ import annotations

from typing import Any, Optional


# ── Type hint extraction ─────────────────────────────────────────────

def _format_range_hint(
    min_val: int | float | None,
    max_val: int | float | None,
    prefix: str,
) -> str:
    if min_val is not None and max_val is not None:
        return f"{{{prefix}:{min_val}}}" if min_val == max_val else f"{{{prefix}:{min_val}-{max_val}}}"
    if min_val is not None:
        return f"{{{prefix}:{min_val}+}}"
    if max_val is not None:
        return f"{{{prefix}:≤{max_val}}}"
    return ""


def _get_length_hint(field_schema: dict[str, Any]) -> str:
    ftype = field_schema.get("type", "string")
    if ftype == "string":
        return _format_range_hint(field_schema.get("minLength"), field_schema.get("maxLength"), "len")
    if ftype in ("integer", "number"):
        return _format_range_hint(field_schema.get("minimum"), field_schema.get("maximum"), "range")
    if ftype == "array":
        return _format_range_hint(field_schema.get("minItems"), field_schema.get("maxItems"), "len")
    return ""


def _parse_metadata_hints(description: str | None) -> str:
    """Extract metadata hints from description like 'format:OCC|pattern:^[A-Z]+$'."""
    if not description:
        return ""
    import re
    match = re.search(r"\(([^)]+)\)", description)
    if not match:
        return ""
    metadata_str = match.group(1)
    hints: list[str] = []
    for part in metadata_str.split("|"):
        pieces = part.split(":", 1)
        if len(pieces) == 2:
            hints.append(f"{{{pieces[0]}:{pieces[1]}}}")
    return "".join(hints)


def _get_type_hint(field_schema: dict[str, Any], include_format: bool = True) -> str:
    """Extract a compact type hint string from a JSON Schema field definition."""
    length_hint = _get_length_hint(field_schema)
    metadata_hints = _parse_metadata_hints(field_schema.get("description"))

    def append_hints(base: str) -> str:
        return base + length_hint + metadata_hints

    # Enum
    if "enum" in field_schema:
        return append_hints("|".join(str(v) for v in field_schema["enum"]))

    # Array
    if field_schema.get("type") == "array":
        items = field_schema.get("items", {})
        if items:
            elem_type = _get_type_hint(items, include_format=False)
            return append_hints(f"array({elem_type})")
        return append_hints("array")

    # Object with properties
    if field_schema.get("type") == "object" and "properties" in field_schema:
        required = set(field_schema.get("required", []))
        fields = []
        for key, prop in field_schema["properties"].items():
            prop_type = _get_type_hint(prop, include_format=False)
            suffix = "" if key in required else "?"
            fields.append(f"{key}{suffix}:{prop_type}")
        return append_hints(f"object({', '.join(fields)})")

    # Object without properties (dict/record)
    if field_schema.get("type") == "object":
        return append_hints("json")

    # Primitives
    type_map = {"string": "string", "number": "number", "integer": "integer", "boolean": "boolean"}
    return append_hints(type_map.get(field_schema.get("type", "string"), "string"))


# ── CSV Schema ────────────────────────────────────────────────────────

def schema_to_csv(
    schema: dict[str, Any],
    *,
    comment: str | None = None,
) -> str:
    """Convert a JSON Schema to a compact CSV header line with type hints.

    Produces output like::

        # Schema: name (string), age (integer)?, role (enum:admin|user)
        name,age,role

    Args:
        schema: A JSON Schema dict (type: object with properties, or type: array with items).
        comment: Optional label for the schema comment line.

    Returns:
        CSV schema string suitable for LLM consumption.
    """
    resolved = _resolve_ref(schema)

    properties: dict[str, Any] = {}
    required: list[str] = []

    if resolved.get("type") == "array" and "items" in resolved:
        properties = resolved["items"].get("properties", {})
        required = resolved["items"].get("required", [])
    elif resolved.get("type") == "object":
        properties = resolved.get("properties", {})
        required = resolved.get("required", [])
    elif "properties" in resolved:
        properties = resolved["properties"]
        required = resolved.get("required", [])

    if not properties:
        raise ValueError(f"Schema must have properties (object or array with items). Got: {list(resolved.keys())}")

    fields: list[str] = []
    type_hints: list[str] = []
    optionals: list[bool] = []

    for field_name, field_schema in properties.items():
        fields.append(field_name)
        is_optional = field_name not in required
        optionals.append(is_optional)
        type_hints.append(_get_type_hint(field_schema))

    # Build schema comment
    schema_parts = [
        f"{fields[i]} ({type_hints[i]}){'?' if optionals[i] else ''}"
        for i in range(len(fields))
    ]
    comment_line = f"# {comment}: " if comment else "# Schema: "
    header_line = ",".join(fields)

    return f"{comment_line}{', '.join(schema_parts)}\n{header_line}"


# ── Markdown Schema ───────────────────────────────────────────────────

def schema_to_markdown(
    schema: dict[str, Any],
    *,
    title: str | None = None,
    description: str | None = None,
) -> str:
    """Convert a JSON Schema to markdown field documentation.

    Produces output like::

        ### MyTool Schema (my_tool)[Get data from the API]

        (This tool uses a method-based API. Specify the 'method' parameter to choose the operation)
        **Fields:**
        - `name` (string): The name of the thing
        - `limit` (integer){range:1-100}?: Max items to return
        - `method` (enum:read|write): Operation to perform

    Args:
        schema: A JSON Schema dict (type: object with properties).
        title: Optional tool name.
        description: Optional tool description.

    Returns:
        Markdown documentation string.
    """
    resolved = _resolve_ref(schema)

    properties = resolved.get("properties", {})
    required = set(resolved.get("required", []))

    lines: list[str] = []

    # Header
    header = "### Tool Schema"
    if title:
        header = f"### {title}"
    if description:
        header += f"[{description}]"
    lines.append(header)

    # Check for method-based API
    method_field = properties.get("method", {})
    if "enum" in method_field:
        lines.append("(This tool uses a method-based API. Specify the 'method' parameter to choose the operation)")

    # Fields
    if properties:
        lines.append("**Fields:**")
        for field_name, field_schema in properties.items():
            type_hint = _get_type_hint(field_schema)
            is_required = field_name in required
            desc = field_schema.get("description", "").strip()

            # Clean up description — strip metadata hints already captured in type
            clean_desc = _parse_metadata_hints(field_schema.get("description"))
            if clean_desc:
                # Remove the (format:...|pattern:...) part from description
                import re
                desc = re.sub(r"\([^)]*(?:format|pattern)[^)]*\)", "", desc).strip()

            suffix = "" if is_required else "?"
            line = f"- `{field_name}` ({type_hint}){suffix}"
            if desc:
                line += f": {desc}"
            lines.append(line)

    return "\n".join(lines)


# ── JSON to CSV data format ──────────────────────────────────────────

def data_to_csv(
    data: list[dict[str, Any]],
    *,
    columns: list[str] | None = None,
    include_header: bool = True,
    fence_block: str | None = None,
) -> str:
    """Convert a list of dicts to CSV format for LLM consumption.

    Args:
        data: List of row dicts.
        columns: Optional column order. Auto-detected from data if omitted.
        include_header: Whether to include a header row.
        fence_block: Optional label to wrap in a ```csv comment block.

    Returns:
        CSV string.
    """
    if not data:
        return ""

    # Determine columns
    if not columns:
        seen: list[str] = []
        for row in data:
            for key in row:
                if key not in seen:
                    seen.append(key)
        columns = seen

    lines: list[str] = []

    if include_header:
        lines.append(",".join(columns))

    for row in data:
        values = [_escape_csv_field(row.get(col, "")) for col in columns]
        lines.append(",".join(values))

    csv_text = "\n".join(lines)

    if fence_block:
        return f"```csv\n# {fence_block}\n{csv_text}\n```"

    return csv_text


# ── Internal helpers ──────────────────────────────────────────────────

def _resolve_ref(schema: dict[str, Any]) -> dict[str, Any]:
    """Resolve $ref references in a JSON Schema."""
    if "$ref" in schema and "definitions" in schema:
        ref_path = schema["$ref"].replace("#/definitions/", "")
        return schema.get("definitions", {}).get(ref_path, schema)
    return schema


def _escape_csv_field(value: Any) -> str:
    """Escape a single value for CSV output."""
    if value is None:
        return ""
    if isinstance(value, bool):
        return str(value).lower()
    if isinstance(value, (list, dict)):
        import json
        s = json.dumps(value, separators=(",", ":"))
        if "," in s or '"' in s or "\n" in s:
            return f'"{s.replace(chr(34), chr(34) + chr(34))}"'
        return s
    s = str(value)
    if "," in s or '"' in s or "\n" in s:
        return f'"{s.replace(chr(34), chr(34) + chr(34))}"'
    return s

File: packages/js/src/schema.ts
/**
 * Schema formatting — convert JSON Schemas to compact LLM-friendly formats.
 *
 * Opt-in: ``npm install toolbox-gateway`` includes these utilities.
 *
 * This module provides two key utilities:
 *
 * 1. ``schemaToCsv`` — Convert a JSON Schema to a compact CSV header with type hints
 * 2. ``schemaToMarkdown`` — Convert a JSON Schema to markdown field documentation
 *
 * Both are designed for the ``toolbox explain`` command to return compact,
 * LLM-friendly tool documentation instead of raw JSON Schema blobs.
 */

// ── Types ─────────────────────────────────────────────────────────────

type JsonSchema = Record<string, unknown>;

// ── Type hint extraction ─────────────────────────────────────────────

function formatRangeHint(
  minVal: number | null | undefined,
  maxVal: number | null | undefined,
  prefix: string,
): string {
  if (minVal != null && maxVal != null) {
    return minVal === maxVal
      ? `{${prefix}:${minVal}}`
      : `{${prefix}:${minVal}-${maxVal}}`;
  }
  if (minVal != null) return `{${prefix}:${minVal}+}`;
  if (maxVal != null) return `{${prefix}:≤${maxVal}}`;
  return "";
}

function getLengthHint(fieldSchema: JsonSchema): string {
  const ftype = fieldSchema.type ?? "string";
  if (ftype === "string") {
    return formatRangeHint(
      fieldSchema.minLength as number | undefined,
      fieldSchema.maxLength as number | undefined,
      "len",
    );
  }
  if (ftype === "integer" || ftype === "number") {
    return formatRangeHint(
      fieldSchema.minimum as number | undefined,
      fieldSchema.maximum as number | undefined,
      "range",
    );
  }
  if (ftype === "array") {
    return formatRangeHint(
      fieldSchema.minItems as number | undefined,
      fieldSchema.maxItems as number | undefined,
      "len",
    );
  }
  return "";
}

function parseMetadataHints(description: string | null | undefined): string {
  /** Extract metadata hints from description like 'format:OCC|pattern:^[A-Z]+$'. */
  if (!description) return "";
  const match = description.match(/\(([^)]+)\)/);
  if (!match) return "";
  const metadataStr = match[1];
  const hints: string[] = [];
  for (const part of metadataStr.split("|")) {
    const pieces = part.split(":", 2);
    if (pieces.length === 2) {
      hints.push(`{${pieces[0]}:${pieces[1]}}`);
    }
  }
  return hints.join("");
}

function getTypeHint(fieldSchema: JsonSchema, includeFormat = true): string {
  /** Extract a compact type hint string from a JSON Schema field definition. */
  const lengthHint = getLengthHint(fieldSchema);
  const metadataHints = parseMetadataHints(fieldSchema.description as string | undefined);

  const appendHints = (base: string): string => base + lengthHint + metadataHints;

  // Enum
  if (fieldSchema.enum) {
    return appendHints((fieldSchema.enum as unknown[]).map(String).join("|"));
  }

  // Array
  if (fieldSchema.type === "array") {
    const items = (fieldSchema.items as JsonSchema) ?? {};
    if (Object.keys(items).length > 0) {
      const elemType = getTypeHint(items, false);
      return appendHints(`array(${elemType})`);
    }
    return appendHints("array");
  }

  // Object with properties
  if (fieldSchema.type === "object" && fieldSchema.properties) {
    const required = new Set(fieldSchema.required as string[] ?? []);
    const props = fieldSchema.properties as Record<string, JsonSchema>;
    const fields: string[] = [];
    for (const [key, prop] of Object.entries(props)) {
      const propType = getTypeHint(prop, false);
      const suffix = required.has(key) ? "" : "?";
      fields.push(`${key}${suffix}:${propType}`);
    }
    return appendHints(`object(${fields.join(", ")})`);
  }

  // Object without properties (dict/record)
  if (fieldSchema.type === "object") {
    return appendHints("json");
  }

  // Primitives
  const typeMap: Record<string, string> = {
    string: "string",
    number: "number",
    integer: "integer",
    boolean: "boolean",
  };
  return appendHints(typeMap[fieldSchema.type as string] ?? "string");
}

// ── CSV Schema ──────────────────────────────────────────────────────

export function schemaToCsv(
  schema: JsonSchema,
  comment?: string,
): string {
  /** Convert a JSON Schema to a compact CSV header line with type hints. */
  const resolved = resolveRef(schema);

  let properties: Record<string, JsonSchema> = {};
  let required: string[] = [];

  if (resolved.type === "array" && resolved.items) {
    const items = resolved.items as JsonSchema;
    properties = (items.properties as Record<string, JsonSchema>) ?? {};
    required = (items.required as string[]) ?? [];
  } else if (resolved.type === "object") {
    properties = (resolved.properties as Record<string, JsonSchema>) ?? {};
    required = (resolved.required as string[]) ?? [];
  } else if (resolved.properties) {
    properties = resolved.properties as Record<string, JsonSchema>;
    required = (resolved.required as string[]) ?? [];
  }

  const propKeys = Object.keys(properties);
  if (propKeys.length === 0) {
    throw new Error(
      `Schema must have properties (object or array with items). Got: ${Object.keys(resolved)}`,
    );
  }

  const fields: string[] = [];
  const typeHints: string[] = [];
  const optionals: boolean[] = [];

  for (const fieldName of propKeys) {
    const fieldSchema = properties[fieldName];
    fields.push(fieldName);
    const isOptional = !required.includes(fieldName);
    optionals.push(isOptional);
    typeHints.push(getTypeHint(fieldSchema));
  }

  // Build schema comment
  const schemaParts = fields.map(
    (f, i) => `${f} (${typeHints[i]})${optionals[i] ? "?" : ""}`,
  );
  const commentLine = comment ? `# ${comment}: ` : "# Schema: ";
  const headerLine = fields.join(",");

  return `${commentLine}${schemaParts.join(", ")}\n${headerLine}`;
}

// ── Markdown Schema ─────────────────────────────────────────────────

export function schemaToMarkdown(
  schema: JsonSchema,
  opts?: { title?: string; description?: string },
): string {
  const title = opts?.title;
  const description = opts?.description;
  /** Convert a JSON Schema to markdown field documentation. */
  const resolved = resolveRef(schema);

  const properties = (resolved.properties as Record<string, JsonSchema>) ?? {};
  const required = new Set(resolved.required as string[] ?? []);

  const lines: string[] = [];

  // Header
  let header = "### Tool Schema";
  if (title) header = `### ${title}`;
  if (description) header += `[${description}]`;
  lines.push(header);

  // Check for method-based API
  const methodField = properties.method ?? {};
  if (methodField.enum) {
    lines.push(
      "(This tool uses a method-based API. Specify the 'method' parameter to choose the operation)",
    );
  }

  // Fields
  if (Object.keys(properties).length > 0) {
    lines.push("**Fields:**");
    for (const [fieldName, fieldSchema] of Object.entries(properties)) {
      const typeHint = getTypeHint(fieldSchema);
      const isRequired = required.has(fieldName);
      let desc = ((fieldSchema.description as string) ?? "").trim();

      // Clean up description — strip metadata hints already captured in type
      const cleanDesc = parseMetadataHints(fieldSchema.description as string | undefined);
      if (cleanDesc) {
        // Remove the (format:...|pattern:...) part from description
        desc = desc.replace(/\([^)]*(?:format|pattern)[^)]*\)/g, "").trim();
      }

      const suffix = isRequired ? "" : "?";
      let line = `- \`${fieldName}\` (${typeHint})${suffix}`;
      if (desc) line += `: ${desc}`;
      lines.push(line);
    }
  }

  return lines.join("\n");
}

// ── JSON to CSV data format ─────────────────────────────────────────

export function dataToCsv(
  data: Record<string, unknown>[],
  opts?: {
    columns?: string[];
    includeHeader?: boolean;
    fenceBlock?: string;
  },
): string {
  /** Convert a list of objects to CSV format for LLM consumption. */
  if (data.length === 0) return "";

  // Determine columns
  let columns = opts?.columns;
  if (!columns) {
    const seen: string[] = [];
    for (const row of data) {
      for (const key of Object.keys(row)) {
        if (!seen.includes(key)) seen.push(key);
      }
    }
    columns = seen;
  }

  const includeHeader = opts?.includeHeader ?? true;
  const lines: string[] = [];

  if (includeHeader) {
    lines.push(columns.join(","));
  }

  for (const row of data) {
    const values = columns.map((col) => escapeCsvField(row[col] ?? ""));
    lines.push(values.join(","));
  }

  const csvText = lines.join("\n");

  if (opts?.fenceBlock) {
    return `\`\`\`csv\n# ${opts.fenceBlock}\n${csvText}\n\`\`\``;
  }

  return csvText;
}

// ── Internal helpers ────────────────────────────────────────────────

function resolveRef(schema: JsonSchema): JsonSchema {
  /** Resolve $ref references in a JSON Schema. */
  if (schema["$ref"] && schema.definitions) {
    const refPath = (schema["$ref"] as string).replace("#/definitions/", "");
    const defs = schema.definitions as Record<string, JsonSchema>;
    return defs[refPath] ?? schema;
  }
  return schema;
}

function escapeCsvField(value: unknown): string {
  /** Escape a single value for CSV output. */
  if (value == null) return "";
  if (typeof value === "boolean") return String(value);
  if (Array.isArray(value) || (typeof value === "object" && value !== null)) {
    const s = JSON.stringify(value);
    if (s.includes(",") || s.includes('"') || s.includes("\n")) {
      return `"${s.replace(/"/g, '""')}"`;
    }
    return s;
  }
  const s = String(value);
  if (s.includes(",") || s.includes('"') || s.includes("\n")) {
    return `"${s.replace(/"/g, '""')}"`;
  }
  return s;
}
